NE-703, Added config details for TLS for forwarded DNS queries (#45373)
openshift/openshift-docs: modules/nw-dns-forward.adoc

$ oc edit dns.operator/default
----
+

> This allows the Operator to create and update the config map named `dns-default` with additional server configuration blocks based on `Server`. If none of the servers has a zone that matches the query, then name resolution falls back to the upstream DNS servers.

Reviewer: ```suggestion
This allows the Operator to create and update the config map named `dns-default` with additional server configuration blocks based on `Server`. If none of the servers have a zone that matches the query, then name resolution falls back to the upstream DNS servers.
```

    - foo.com
    forwardPlugin:
      policy: Random <3>
      transportConfig:
        tls:

>           caBundle

Reviewer 2: Found one minor thing here; `caBundle` should be `caBundle:` (with a semi-colon)

            name: mycacert

>           serverName: upstream-tls <4>

Reviewer 2: Since most cases we'll expect an FQDN, it might make sense to expand this a little more to be something like `dnstls.example.com`.

      upstreams:
      - 3.3.3.3
      - 4.4.4.4:5454
  upstreamResolvers: <7>
    policy: Random <8>
    upstreams: <9>
    - type: SystemResolvConf <10>
    - type: Network
      address: 1.2.3.4 <11>
      port: 53 <12>
----
<1> Must comply with the `rfc6335` service name syntax.
<2> Must conform to the definition of a `subdomain` in `rfc1123`. The cluster domain, `cluster.local`, is an invalid `subdomain` for `zones`.
<3> Defines the policy to select upstream resolvers. Default value is `Random`. You can also use `RoundRobin`, and `Sequential`.

> <4> If you are configuring TLS for forwarded DNS queries, this is a mandatory server name used as part of the server name indication (SNI) to validate the upstream.

Reviewer 2: ```suggestion
<4> If you are configuring TLS for forwarded DNS queries, this is a mandatory server name used as part of the server name indication (SNI) to validate the upstream TLS server certificate.
```

<7> Optional. You can use it to provide upstream resolvers.
<8> You can specify two types of `upstreams` - `SystemResolvConf` and `Network`. `SystemResolvConf` configures the upstream to use `/etc/resolv.conf` and `Network` defines a `Networkresolver`. You can specify one or both.
<9> If the specified type is `Network`, you must provide an IP address. `address` must be a valid IPv4 or IPv6 address.

> <10> If the specified type is `Network`, you can optionally provide a port. `port` must be between 1 and 65535. If you do not specify a port for the upstream, by default port 853 is tried.

Reviewer: Suggestion to specify [what`port` is](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-object-references):
```suggestion
<10> If the specified type is `Network`, you can optionally provide a port. The `port` field must have a value between `1` and `65535`. If you do not specify a port for the upstream, by default port 853 is tried.
```
since the numbers mentioned are potential values for that field, my thought is that they need to be wrapped in backticks as well. But feel free to disregard if not 🙂

Author: Great catch!

+
When working in a highly regulated environment, you might need the ability to secure DNS traffic when forwarding requests to upstream resolvers so that you can ensure additional DNS traffic and data privacy. Cluster administrators can configure transport layer security (TLS) for forwarded DNS queries.
+
.Configuring DNS forwarding with TLS
[source,yaml]
----
apiVersion: operator.openshift.io/v1
kind: DNS
metadata:
  name: default
spec:
  servers:
  - name: example-server <1>
    zones: <2>
    - example.com
    forwardPlugin:
      transportConfig:
        transport: TLS <3>
        tls:
          caBundle:
            name: mycacert
          serverName: dnstls.example.com  <4>
      policy: Random <5>
      upstreams: <6>
      - 1.1.1.1
      - 2.2.2.2:5353
  upstreamResolvers: <7>
    transportConfig:
      transport: TLS
      tls:
        caBundle:
          name: mycacert
        serverName: dnstls.example.com
    upstreams:
    - type: Network <8>
      address: 1.2.3.4 <9>
      port: 53 <10>
----
<1> Must comply with the `rfc6335` service name syntax.

> <2> Must conform to the definition of a `subdomain` in `rfc1123`. The cluster domain, `cluster.local`, is an invalid `subdomain` for `zones`.

Reviewer: Same kind of object reference suggestion
```suggestion
<2> Must conform to the definition of a subdomain in the `rfc1123` service name syntax. The cluster domain, `cluster.local`, is an invalid subdomain for the `zones` field.
```

> <3> When configuring TLS for forwarded DNS queries, set `transport` to `TLS`.

Reviewer: ```suggestion
<3> When configuring TLS for forwarded DNS queries, set the `transport` field to have the value `TLS`.
```

By default, CoreDNS caches forwarded connections for 10 seconds. CoreDNS will hold a TCP connection open for those 10 seconds if no request is issued. With large clusters, ensure that your DNS server is aware that it might get many new connections to hold open because you can initiate a connection per node. Set up your DNS hierarchy accordingly to avoid performance issues.
<4> When configuring TLS for forwarded DNS queries, this is a mandatory server name used as part of the server name indication (SNI) to validate the upstream TLS server certificate.

> <5> Defines the policy to select upstream resolvers. Default value is `Random`. You can also use `RoundRobin`, and `Sequential`.

Reviewer: ```suggestion
<5> Defines the policy to select upstream resolvers. Default value is `Random`. You can also use the values `RoundRobin`, and `Sequential`.
```

> <6> Required. You can use it to provide upstream resolvers. A maximum of 15 `upstreams` is allowed per `forwardPlugin`.

Reviewer: ```suggestion
<6> Required. You can use it to provide upstream resolvers. A maximum of 15 `upstreams` entries are allowed per `forwardPlugin` entry.
```

<7> Optional. You can use it to override the default policy and forward DNS resolution to the specified DNS resolvers (upstream resolvers) for the default domain. If you do not provide any upstream resolvers, the DNS name queries go to the servers in `/etc/resolv.conf`.

> <8>`Network` defines a `Networkresolver`. Only the `Network` type is allowed when using TLS and you must provide an IP address.

Reviewer 2: ```suggestion
<8>`Network` type indicates that this upstream resolver should handle forwarded requests separately from the upstream resolvers listed in `/etc/resolv.conf`. Only the `Network` type is allowed when using TLS and you must provide an IP address.
```

> <9> `address` must be a valid IPv4 or IPv6 address.

Reviewer: ```suggestion
<9> The `address` field must be a valid IPv4 or IPv6 address.
```

> <10> You can optionally provide a port. `port` must be between 1 and 65535. If you do not specify a port for the upstream, by default port 853 is tried.

Reviewer: ```suggestion
<10> You can optionally provide a port. The `port` field must have a value between `1` and `65535`. If you do not specify a port for the upstream, by default port 853 is tried.
```
